### DISCUSSION TITLE
Unexpected behavior of no_args_is_help with 0 or 1 commands

### DISCUSSION THREAD

**Discussion by jacob-bush-shopify** [fastapi/typer#937] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Typer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Typer" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to Typer but to [Click](https://github.com/pallets/click).

#### Example 1
I expect help when invoked with no args
```python
import typer

cli = typer.Typer(no_args_is_help=True)

if __name__ == "__main__":
    cli()
```
Actually raises an exception
```
$ python cli.py
AssertionError: Could not get a command for this Typer instance
```

#### Example 2
I expect help when invoked with no args
```python
import typer

cli = typer.Typer(no_args_is_help=True)

@cli.command("1") 
def _1():
    print(1)

if __name__ == "__main__":
    cli()
```
Actually prints "1"
```
$ python cli.py
1
```
#### Example 3
I expect help when invoked with no args
```python
import typer

cli = typer.Typer(no_args_is_help=True)

@cli.command("1") 
def _1():
    print(1)

@cli.command("2") 
def _2():
    print(2)

if __name__ == "__main__":
    cli()
```
Works as expected (though this breaks code relying on the behaviour of the 1-command typer app)
```
$ python cli.py
Usage: cli.py [OPTIONS] COMMAND [ARGS]...

Options:
  --install-completion [bash|zsh|fish|powershell|pwsh]
                                  Install completion for the specified shell.
  --show-completion [bash|zsh|fish|powershell|pwsh]
                                  Show completion for the specified shell, to
                                  copy it or customize the installation.
  --help                          Show this message and exit.

Commands:
  1
  2
```

### Description
When `no_args_is_help=True` is set on my Typer application, I expect that the application will print the help text when it is called with no arguments.

When 2+ commands are registered, `no_args_is_help` works as expected.

However, there are two cases where this does not hold:
- When there are 0 commands registered on the application, an exception is raised
- When there is 1 command registered on the application, the command is called in lieu of printing the help text.

### Operating System

macOS

### Typer Version

0.6.1

### Python Version

3.9.12

**Comment by marcstreeter** [comment]
I solved this for myself with [this gist](https://gist.github.com/marcstreeter/2eaa8ae94c58f953ce7df852ee8d846c).

**Comment by tiangolo** [comment, marked as answer]
Thanks for the report!

Let's think about each case individually.

## 2 commands

So, the case with 2 commands is working as expected, right?

## 1 command

For the case of one command, it's currently working as intended, at least that's how it was designed from the beginning. When there's an app with a single command, that command is called directly, there's no subcommand created for it. It's documented here: https://typer.tiangolo.com/tutorial/commands/#explicit-application

If you want it to be a subcommand you would need to do one of:

* Add another command (which you wouldn't want in this case)
* Add a callback, it can be just a no-op, but it would tell Typer that this app should be considered a "group" with subcommands, even if just one.

```Python
import typer

cli = typer.Typer(no_args_is_help=True)

@cli.command("1") 
def _1():
    print(1)

@cli.callback()
def callback():
    pass

if __name__ == "__main__":
    cli()
```

Then:

```console
$ python demo.py  
                                                                            
 Usage: demo.py [OPTIONS] COMMAND [ARGS]...                                 
                                                                            
╭─ Options ────────────────────────────────────────────────────────────────╮
│ --install-completion          Install completion for the current shell.  │
│ --show-completion             Show completion for the current shell, to  │
│                               copy it or customize the installation.     │
│ --help                        Show this message and exit.                │
╰──────────────────────────────────────────────────────────────────────────╯
╭─ Commands ───────────────────────────────────────────────────────────────╮
│ 1                                                                        │
╰──────────────────────────────────────────────────────────────────────────╯
```

## no command

I would actually consider an app with no command or callback to be an error, telling me that there's no command to run, not a help message telling me that the only thing I can do is get that help message. :thinking:

**Comment by imLew** [comment]
I was also surprised by the one command behavior and adding another command would be confused for users imo.

It'd be a neat feature to have as it would save users just a little bit of frustration.
Edit: It'd be great if this argument could be given to `typer.run` for single command apps.
